forge/src/trace/decoder.rs

```diff
@@ -0,0 +1,234 @@
+use super::{
+    CallTraceArena, RawOrDecodedCall, RawOrDecodedLog, RawOrDecodedReturnData, TraceIdentifier,
+};
+use crate::abi::{CHEATCODE_ADDRESS, CONSOLE_ABI, HEVM_ABI};
+use ethers::{
+    abi::{Abi, Address, Event, Function, Token},
+    types::H256,
+};
+use foundry_utils::format_token;
+use std::collections::BTreeMap;
+
+/// The call trace decoder.
+///
+/// The decoder collects address labels and ABIs from any number of [TraceIdentifier]s, which it
+/// then uses to decode the call trace.
+///
+/// Note that a call trace decoder is required for each new set of traces, since addresses in
+/// different sets might overlap.
+#[derive(Default, Debug)]
+pub struct CallTraceDecoder {
+    /// Addresses identified to be a specific contract.
+    ///
+    /// The values are in the form `"<artifact>:<contract>"`.
+    pub contracts: BTreeMap<Address, String>,
+    /// Address labels
+    pub labels: BTreeMap<Address, String>,
+    /// All known functions
+    pub functions: BTreeMap<[u8; 4], Function>,
+    /// All known events
+    pub events: BTreeMap<H256, Event>,
+    /// All known errors
+    pub errors: Abi,
+}
+
+impl CallTraceDecoder {
+    /// Creates a new call trace decoder.
+    ///
+    /// The call trace decoder always knows how to decode calls to the cheatcode address, as well
+    /// as DSTest-style logs.
+    pub fn new() -> Self {
+        Self {
+            contracts: BTreeMap::new(),
+            labels: [(*CHEATCODE_ADDRESS, "VM".to_string())].into(),
```

Comment:
agree that would be nice - tho we should think about how we want to handle L2 precompiles. I would like to get to the point of easily supporting a different set of precompiles for various chains & L2s

Reply:
Good point on the L2 precompiles. I created an issue here where we can discuss instead since I think we need to give it more thought: https://github.com/gakonst/foundry/issues/916